Libraries/LLM/Phi.swift

```diff
@@ -84,15 +78,13 @@ private class PhiAttention: Module {
 
         // Finally perform the attention computation
         let scale = sqrt(1 / Float(queries.dim(-1)))
-        var scores = (queries * scale).matmul(keys.transposed(0, 1, 3, 2))
-        if let mask {
-            scores = scores + mask
-        }
-
-        scores = softMax(scores, axis: -1).asType(values.dtype)
-        let valuesHat = matmul(scores, values).transposed(0, 2, 1, 3).reshaped(B, L, -1)
+        let output = MLXFast.scaledDotProductAttention(
+            queries: queries, keys: keys, values: values, scale: scale, mask: mask
+        )
+        .transposed(0, 2, 1, 3)
+        .reshaped(B, L, -1)
```

Comment:
It just occurred to me there is an issue here? I think the output will be in fp32 since you are sending the inputs in fp32 (that will propagate through the rest of the model and slow things way down). We need to explicitly downcast the output after the SDPA (might be good to add a test for that).

Reply:
ah perhaps -- I guess it would still run, so I didn't notice.

Indeed it now produces float32 instead of float16.  So something like this?

```
        let output = MLXFast.scaledDotProductAttention(
            queries: queries, keys: keys, values: values, scale: scale, mask: mask
        )
        .asType(values.dtype)
        .transposed(0, 2, 1, 3)
        .reshaped(B, L, -1)
```

Reply:
Yea I would make a couple changes here:
1. Cast the output to the values type (as suggested)
2. Don't upcast the keys, because we don't want to store fp32 in the KV cache. Just upcast the queries and let the op handle type promotion.